beacon-chain/sync/receive_block.go

```diff
@@ -0,0 +1,148 @@
+		trace.Int64Attribute("finalized slot", int64(beaconState.FinalizedEpoch*params.BeaconConfig().SlotsPerEpoch)),
+	)
+	if block.Slot < beaconState.FinalizedEpoch*params.BeaconConfig().SlotsPerEpoch {
+		log.Debug("Discarding received block with a slot number smaller than the last finalized slot")
+		return nil
+	}
+
+	// We check if we have the block's parents saved locally.
+	parentRoot := bytesutil.ToBytes32(block.ParentRootHash32)
+	hasParent := rs.db.HasBlock(parentRoot)
+	span.AddAttributes(trace.BoolAttribute("hasParent", hasParent))
+
+	if !hasParent {
+		// If we do not have the parent, we insert it into a pending block's map.
+		rs.insertPendingBlock(ctx, blockRoot, msg)
+		// We update the last observed slot to the received canonical block's slot.
+		if block.Slot > rs.highestObservedSlot {
+			rs.highestObservedSlot = block.Slot
+		}
+		return nil
+	}
+
+	log.WithField("blockRoot", fmt.Sprintf("%#x", blockRoot)).Debug("Sending newly received block to chain service")
+	// We then process the block by passing it through the ChainService and running
+	// a fork choice rule.
+	beaconState, err = rs.chainService.ReceiveBlock(ctx, block)
+	if err != nil {
+		log.Errorf("Could not process beacon block: %v", err)
+		return err
+	}
+	if err := rs.chainService.ApplyForkChoiceRule(ctx, block, beaconState); err != nil {
+		log.Errorf("could not apply fork choice rule: %v", err)
+		return err
+	}
+
+	// We clear the block root from the pending processing map.
+	rs.clearPendingBlock(blockRoot)
+	sentBlocks.Inc()
+
+	// We update the last observed slot to the received canonical block's slot.
+	if block.Slot > rs.highestObservedSlot {
+		rs.highestObservedSlot = block.Slot
+	}
+	span.AddAttributes(trace.Int64Attribute("highestObservedSlot", int64(rs.highestObservedSlot)))
+
+	// If the block has a child, we then clear it from the blocks pending processing
+	// and call receiveBlock recursively. The recursive function call will stop once
+	// the block we process no longer has children.
+	if child, ok := rs.hasChild(blockRoot); ok {
+		return rs.receiveBlock(child)
+	}
+	return nil
+}
+
+func (rs *RegularSync) insertPendingBlock(ctx context.Context, blockRoot [32]byte, blockMsg p2p.Message) {
+	rs.blocksAwaitingProcessingLock.Lock()
+	defer rs.blocksAwaitingProcessingLock.Unlock()
+	rs.blocksAwaitingProcessing[blockRoot] = blockMsg
+	blocksAwaitingProcessingGauge.Inc()
+	rs.p2p.Broadcast(ctx, &pb.BeaconBlockRequest{Hash: blockRoot[:]})
+}
+
+func (rs *RegularSync) clearPendingBlock(blockRoot [32]byte) {
+	rs.blocksAwaitingProcessingLock.Lock()
+	defer rs.blocksAwaitingProcessingLock.Unlock()
+	delete(rs.blocksAwaitingProcessing, blockRoot)
+	blocksAwaitingProcessingGauge.Dec()
+}
+
+func (rs *RegularSync) hasChild(blockRoot [32]byte) (p2p.Message, bool) {
+	rs.blocksAwaitingProcessingLock.Lock()
+	defer rs.blocksAwaitingProcessingLock.Unlock()
+	child, ok := rs.blocksAwaitingProcessing[blockRoot]
```

Comment:
```go
return rs.blocksAwaitingProcessing[blockRoot]
```

Reply:
Maps are weird in the sense that if we do that, it only returns the child and not the bool ok value. We have to unpack it in order to extract both @prestonvanloon.

Reply:
got it